### ISSUE TITLE
Fail safely if tuple limit is reached

### ISSUE THREAD

**Issue by kyukhin** [tarantool/tarantool#3594]
If tuple limit (which is around 2B right now) is reached, Tarantool segfaults like this:
```
0  0x00000000004787e3 in matras_view_get_no_check (v=0x227b688, id=4294967295, m=0x227b688) at /var/tmp/portage/dev-db/tarantool-1.6.9-r9999/work/tarantool-1.6.9.80/src/lib/small/small/matras.h:353
#1  matras_get (id=4294967295, m=0x227b688) at /var/tmp/portage/dev-db/tarantool-1.6.9-r9999/work/tarantool-1.6.9.80/src/lib/small/small/matras.h:372
#2  light_index_insert (value=<optimized out>, hash=1314228454, ht=0x227b670) at /var/tmp/portage/dev-db/tarantool-1.6.9-r9999/work/tarantool-1.6.9.80/src/lib/salad/light.h:786
```

Need to reports a error and exit normally.

**Comment by kshcherbatov** [comment]
Ncurses on Fedora 28 is compiled with flag  
--fstack-clash-protection that use stack protection 
mechanism(strictly speaking configure option --enable-widec 
is also required, but it is not a part of problem we
investigated): gcc inserting code to step the stack down 
one page at a time, running a logical-OR with zero 
at each point, which doesn't affect any value on the 
stack but forces a memory access:
```
  lea r11,[rsp-frameSize]
label:
  sub rsp,pageSize
  or QWORD PTR [rsp],0x0
  cmp rsp,r11
  jne label
```
where frameSize=32768 b and pageSize=4096 b
(read also https://ldpreload.com/blog/stack-smashes-you)

Tarantool main interactive loop is working in fiber with
default stack size 65536 b
```
BINARY IMAGE MEMORY MAP:

SECTION      ADDRESSES       COMMENT
             
DATA         0x0       ^
HEAP                   |
             0x0ec18   |  # < --fstack-clash-
                       |  #   protection check
STACK: @               |  #
       @     0x16c18   | $# < ncurses/readline 
       @               | $    internals, access
       @               | $    syscall in wrapper
       @               | $
       @               | $
       @     0x1ffe0   | $  < frame0 -- LUA
             ....      |
             0xffff    |
             
+------+-----------------------------+-------------+
| SIGN | DESCRIPTION                 | TOTAL SIZE  |
+------+-----------------------------+-------------+
|  @   | stack area region; (fiber)  |   65536 b   |
+------+-----------------------------+-------------+
|  $   | user-space application stack|   37832 b   |
|      | memory usage                |             |
+------+-----------------------------+-------------+
|  #   | a memory that checked stack |   32768 b   |
|      | probing generated with      |             |
|      | --fstack-clash-protection   |             |
+------+-----------------------------+-------------+
```
in other words, $ + # = 70600    >    65536 = @
and we have segfault
```
SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_ACCERR, si_addr=}
SIGSEGV {si_signo=SIGSEGV, si_code=SI_KERNEL, si_addr=NULL}
```

So, I suppose to increase Tarantool interactive loop fiber stack
to the size of 131072 b. Program would stop crashing on launch
and init-scripts would also works.

**Comment by Totktonada** [comment]
Let's investigate and update https://github.com/tarantool/doc/issues/586 (and https://github.com/tarantool/doc/pull/1033) with actual limit.

**Comment by Korablev77** [comment]
Seems like it concerns only hash indexes: each bucket contains (at most) one tuple and bucket itself is allocated using matras. In btree only blocks are allocated using matras allocator. In turn each (inner) block comprises up to 24 elements which allows it to fit way more than 2^32 tuples. Fix me if I am wrong.

**Comment by Gumix** [comment]
It seems that currently there are no 2B or 4B limits for the hash index.
I managed to insert 5.48B tuples until Tarantool was killed by OOM.

```lua
box.cfg{wal_mode = 'none', memtx_memory = 190 * 1024 * 1024 * 1024}
require('fiber').set_max_slice(60)

local s = box.schema.space.create('test')
s:create_index('pk', {type = 'hash'})

local i = 1
while true do
    if i % 1000 == 0 then
        require('fiber').yield()
    end
    if i % 1000000 == 0 then
        print(i / 1000000 .. "M")
    end

    s:insert{i}
    i = i + 1
end
```

**Comment by Gumix** [comment]
> I managed to insert 5.48B tuples into the hash index until Tarantool was killed by OOM.
> 
> ```lua
> box.cfg{wal_mode = 'none', memtx_memory = 190 * 1024 * 1024 * 1024}
> require('fiber').set_max_slice(60)
> 
> local s = box.schema.space.create('test')
> s:create_index('pk', {type = 'hash'})
> 
> local i = 1
> while true do
>     if i % 1000 == 0 then
>         require('fiber').yield()
>     end
>     if i % 1000000 == 0 then
>         print(i / 1000000 .. "M")
>     end
> 
>     s:insert{i}
>     i = i + 1
> end
> ```
> 
> However the actual limit of `matras` is 2^32 (~4B), that means that `light` records with ids > 4B overwrite the records starting from 0.

The output with the fix applied:
```
[...]
4290M
4291M
4292M
4293M
4294M
2023-10-17 23:32:10.984 [200256] main memtx_hash.cc:381 E> OutOfMemory: Failed to allocate 4294967289 bytes in hash_table for key
2023-10-17 23:32:10.994 [200256] main F> fatal error, exiting the event loop
```

BTW, the error message itself it incorrect: 4294967289 is the number of records stored in a light hash table, not the number of bytes.
